lib/Runtime/Debug/DiagObjectModel.cpp

```diff
@@ -285,6 +285,52 @@ namespace Js
         pResolvedObject->objectDisplay->SetDefaultTypeAttribute(defaultAttributes);
     }
 
+    // The debugger uses the functionNameId field instead of the "name" property to get the name of the funtion. The functionNameId field is overloaded and may contain the display name if
+    // toString() has been called on the function object. For built-in or external functions the display name can be something like "function Echo() { native code }". We will try to parse the
+    // function name out of the display name so the user will see just the function name e.g. "Echo" instead of the full display name in debugger.
+    JavascriptString * VariableWalkerBase::ParseFunctionName(JavascriptString* displayName, ScriptContext* scriptContext)
+    {
+        Assert(displayName);
+        const char16 * displayNameBuffer = displayName->GetString();
+        const charcount_t displayNameBufferLength = displayName->GetLength();
+        const charcount_t funcStringLength = _countof(JS_DISPLAY_STRING_FUNCTION_HEADER) - 1; // discount the ending null character in string literal
+        const charcount_t templateStringLength = funcStringLength + _countof(JS_DISPLAY_STRING_FUNCTION_BODY) - 1; // discount the ending null character in string literal
+        // If the string doesn't meet our expected format; return the original string.
+        if (displayNameBufferLength < templateStringLength || (wmemcmp(displayNameBuffer, JS_DISPLAY_STRING_FUNCTION_HEADER, funcStringLength) != 0))
+        {
+            return displayName;
+        }
+
+        // Look for the left parenthesis, if we don't find one; return the original string. If we find a left parenthesis, match the rest of the function body pattern, if we don't find an exact match; return the original string.
+        const char16* parenChar = wmemchr(displayNameBuffer, '(', displayNameBufferLength);
+        if ((parenChar == nullptr) || (wmemcmp(parenChar, JS_DISPLAY_STRING_FUNCTION_BODY, _countof(JS_DISPLAY_STRING_FUNCTION_BODY)) != 0))
+        {
+            return displayName;
+        }
+
+        charcount_t actualFunctionNameLength = displayNameBufferLength - templateStringLength;
+        if (actualFunctionNameLength == 0)
```

Comment:
I had it like that before but Doug pointed out a spec consideration where this could sometimes be "function () { [native code] }". In this case the function name is blank. This happen for built-in or external function from what I have seen so far. But I have handled that case in case we see something like that in future.

---
In reply to: [129432508](https://github.com/Microsoft/ChakraCore/pull/3429#discussion_r129432508) [](ancestors = 129432508)